### ISSUE TITLE
moveRel2, which is apparently supposed to allow for sub-pixel movement, doesn't

### ISSUE THREAD

**Issue by JibbSmart** [mfosse/JoyCon-Driver#71]
I modified it for my purposes, but yeah, I was making small movements with the gyro, the mouse wasn't moving unless I made faster movements, I thought there was a threshold somewhere and couldn't find it, but then I realised that sub-pixel movements weren't working, which is why it seemed like a threshold was there. These are the changes I made, though I've no doubt you know what you're doing:
```
float extraX = 0;
float extraY = 0;

// move relative:
// allow for sub pixel mouse movements
void moveRel2(float x, float y/*, float thresholdX = 0.02, float thresholdY = 0.02*/) {

	int fx = (int)x;
	int fy = (int)y;

	extraX += x - fx;
	extraY += y - fy;

	int iExtraX = (int)extraX;
	int iExtraY = (int)extraY;

	fx += iExtraX;
	fy += iExtraY;

	extraX -= iExtraX;
	extraY -= iExtraY;

	//...
}
```
That's from MouseController.hpp, btw.

**Comment by mfosse** [comment]
Briefly read your email, will make a full reply when I get to class

**Comment by mfosse** [comment]
So I haven't looked at the code for this in a while to see why it's probably broken but I'll try to explain what it does and why it's there.

So in Windows you can't move a mouse less than a pixel, which is problematic for what we need.
So why do we need sub-pixel movement? because if you just take the raw gyro data and plug it into the mouse controller you get really inaccurate and way too sensitive movement, and if you just divide by some number the mouse won't move at all (unless you move the device quickly) (because anything less than 1 isn't registered)

Previously (before gyro calibration was taken into account) if you just add up the sub pixel movements and then moved the mouse when they were greater than 1, you got *really* jittery movement and it was impossible to keep the mouse still, so I made a decay (multiplied the sub-pixel movements by like 0.8 so it would converge to 0 if they didn't add up fast enough)

I don't know if this is still the case since I've taken into account the gyro calibration & I'll look into what you've said about the gyro calibration, but I mostly think what you found was a coincidence / your joycons just need to be calibrated

I'm also not sure what you mean when you say that you had a bias/offset to account for, how are you measuring this? through the debug output?
If your mouse moves when you leave the joycon completely still your joycons either aren't calibrated or there's something very wrong lol

**Comment by mfosse** [comment]
Looking at the code I see that it doesn't do that, it's broken for sure
I think I broke it intentionally a while ago and forgot to fix it :P

**Comment by JibbSmart** [comment]
Right, sorry, the email was only tangentially related to the issue. The issue was just that I wasn't getting sub-pixel movements -- I understand Windows can't do sub-pixel movements, but it didn't seem to be remembering the remainder when rounding to ints, so I made it remember the remainder and accumulate.

Regarding your last paragraph, from what I've read, gyros often have a non-zero at-rest reading, because their calibration is affected by temperature (and perhaps other factors I can't remember). So they can't necessarily be calibrated once and left at that (although a single calibration should usually last for at least a few play sessions). This (from what I've read) is referred to as bias or offset. This can be accounted for by getting their average velocity in each axis when at rest, and then subtracting that from subsequent readings. I thought the factory calibration data might be such a calibration (with user calibration space for subsequent calibrations).

**Comment by mfosse** [comment]
This was definitely a bug which I'll fix when I get home, as for the other thing, make an issue, it's not a quick fix like this is

what it sounds like you want is a second calibration to be done at the start of the program?
as it is now, and from what I understand, what the switch does as well, is use the user calibration if it exists, otherwise default to the factory calibration, which is accurate enough

I could put in a second calibration offset by taking the average over a second or two at the start of the program but I'd have to instruct the user to lay the joycons / controller flat / still when it starts

TLDR: It could probably be done but it's a lot of extra work I don't want to do for (imo) minimal gain, when you can just use the user calibration feature on the switch

Something to note:
I think @CTCaer 's joycon tool allows you to edit the user calibration manually
https://github.com/CTCaer/jc_toolkit/

**Comment by JibbSmart** [comment]
Thanks.

Regarding the calibration thing, I haven't made an issue because I don't know if there's actually a problem. I don't think we're on the same page as to how the calibration info is supposed to be used, and while for now I'll defer to your experience on the matter, as a programmer myself I'm going to spend more time trying to understand the calibration situation better.

Thanks for the link to @CTCaer's tool. That looks really helpful for my own education!

**Comment by JibbSmart** [comment]
Just one more question on the calibration side of things: I can see where sensor_cal is set first to factory calibration, then user calibration if that's present, but where in the code does it actually offset the gyro and accel data with sensor_cal? I've been unable to find it.

**Comment by mfosse** [comment]
sensor_cal isn't used directly see init_bt() in Joycon.hpp
acc_cal_coeff and gyro_cal_coeff are used in the final calculation

**Comment by JibbSmart** [comment]
So it's a multiplier rather than an offset. That's interesting. Thanks.

**Comment by mfosse** [comment]
this might be helpful aswell:
https://github.com/dekuNukem/Nintendo_Switch_Reverse_Engineering/blob/master/imu_sensor_notes.md
